conda_package_streaming/conda_fmt.py

```diff
@@ -0,0 +1,179 @@
+"""
+Tools for creating ``.conda``-format archives.
+
+Uses ``tempfile.SpooledTemporaryFile`` to buffer ``pkg-*.tar`` and
+``info-*.tar``, then compress directly into an open `ZipFile` at the end.
+`SpooledTemporaryFile` buffers the first 10MB of the package and its metadata in
+memory, but writes out to disk for larger packages.
+
+Uses more disk space than ``conda-package-handling`` (temporary uncompressed
+tarballs of the package contents) but accepts streams instead of just
+files-on-the-filesystem.
+"""
+
+from __future__ import annotations
+
+import json
+import shutil
+import tarfile
+import tempfile
+import zipfile
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Callable, Iterator
+
+import zstandard
+
+# increase to reduce speed and increase compression (levels above 19 use much
+# more memory)
+ZSTD_COMPRESS_LEVEL = 19
+# increase to reduce compression and increase speed
+ZSTD_COMPRESS_THREADS = 1
+
+CONDA_PACKAGE_FORMAT_VERSION = 2
+
+# Account for growth from "2 GB of /dev/urandom" to not exceed ZIP64_LIMIT after
+# compression
+CONDA_ZIP64_LIMIT = zipfile.ZIP64_LIMIT - (1 << 18) - 1
+
+
+def anonymize(tarinfo: tarfile.TarInfo):
+    """
+    Pass to ``tarfile.add(..., filter=anonymize)`` to anonymize uid/gid.
+
+    Does not anonymize mtime or any other field.
+    """
+    tarinfo.uid = tarinfo.gid = 0
+    tarinfo.uname = tarinfo.gname = ""
+    return tarinfo
+
+
+class CondaTarFile(tarfile.TarFile):
+    """
+    Subclass of :external+python:py:class:`tarfile.TarFile` that adds members to
+    a second ``info`` tar if they match ``is_info(name)``.
+
+    Create this with ``conda_builder(...)`` which sets up the component
+    archives, then wraps them into a ``.conda`` on exit.
+
+    Only useful for creating, not extracting ``.conda``.
+    """
+
+    info_tar: tarfile.TarFile
+    is_info: Callable
+
+    def __init__(
+        self,
+        *args,
+        info_tar: tarfile.TarFile,
+        is_info=lambda name: name.startswith("info/"),
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.info_tar = info_tar
+        self.is_info = is_info
+
+    def addfile(self, tarinfo, fileobj=None):
+        """
+        Add the TarInfo object ``tarinfo`` to the archive. If ``fileobj`` is
+        given, it should be a binary file, and tarinfo.size bytes are read from
+        it and added to the archive. You can create TarInfo objects directly, or
+        by using ``gettarinfo()``.
+
+        If ``self.is_info(tarinfo.name)`` returns ``True``, add ``tarinfo`` to
+        ``self.info_tar`` instead.
+        """
+        if self.is_info(tarinfo.name):
+            return self.info_tar.addfile(tarinfo, fileobj=fileobj)
+        else:
+            return super().addfile(tarinfo, fileobj)
+
+
+@contextmanager
+def conda_builder(
+    file_id,
```

Comment:
```suggestion
    stem,
```

In modern Python lingo I've seen "stem" being used for this part of a path

Reply:
This term is also [copied from conda-package-handling](https://github.com/conda/conda-package-handling/blob/main/src/conda_package_handling/conda_fmt.py#L116) and is jammed in the middle of `pkg-{name}.tar.zst` e.g. but I see we could change it without obsoleting [-streaming's documentation.](https://conda.github.io/conda-package-streaming/transmute.html)